### ISSUE TITLE
Make HttpServletRequest.newPushBuilder() available for JAX-RS

### ISSUE THREAD

**Issue by bmhm** [OpenLiberty/open-liberty#14336]
**Is your feature request related to a problem? Please describe.**

Unexpected / undocumented  behaviour with http/2 and JAX-RS.

Dear IBM team,

I recently read your blog post about using Servlet-4.0 and `HttpServletRequest.newPushBuilder()`:
https://developer.ibm.com/technologies/java/tutorials/j-javaee8-servlet4/#

However, when I use this in an JAX-RS endpoint, the `PushBuilder` is  always `null`:

```java
@Path("/")
public class MyResource {

  @GET
  @Path("/")
  @Produces(MediaType.APPLICATION_JSON)
  public Response doGetResource(
    @Context final HttpServletRequest request
  ) {
    // always null
    final PushBuilder pushBuilder = request.newPushBuilder();

    // … 

    return Response.status(Status.OK).build();
  }
}
```

**Describe the solution you'd like**

I would like to use the push builder as described in the blog post with JAX-RS endpoints.
I believe the API call I created above should work.

**Describe alternatives you've considered**

No viable alternatives known, as you cannot just refactor your JAX-RS endpoints into servlets where the PushBuilder would not be `null`.

**Additional context**

Tested with Liberty 20.0.0.7 and openJ9 v11.x and 14.x

There is no explicit documentation available. Neither if this should have worked or should not have worked in the first place. All blog posts are usually only about plain Servlets, JSF- and JSP-endpoints.

**Comment by bmhm** [comment]
Nevermind, I was using insomnia to test http/2 push. Requesting the REST endpoint using a browser works just fine.

**Comment by andymc12** [comment]
Right, JAX-RS 2.1 does not officially support (or probably could be better phrased as, does not take advantage of) HTTP/2.  

JAX-RS 2.1 does support Server Sent Events ( https://github.com/OpenLiberty/sample-jaxrs-sse ) - part of HTML5 that enables the server to push data to the clients - I don't think it is as efficient as HTTP/2 Push, but it might suit your needs too.

HTH

**Comment by bmhm** [comment]
> Right, JAX-RS 2.1 does not officially support HTTP/2.

Well, the example from above actually *does* work just fine and I did create some push events through this.

Actually, JAX-RS endpoints are just specialized servlets – which is why I wondered it didn’t work at first.

But then I realized the IBM Servlet implementation has support for this:
https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/srt/SRTServletRequest40.java#L217-L252

This got me thinking to see what happens when I [enable servlet tracing like so](https://www.ibm.com/support/pages/set-trace-and-get-full-dump-websphere-liberty):
(I enabled the servlet part)
```xml
     
<logging
 
 traceSpecification="com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HTTPChannel=all:GenericBNF=all:HTTPDispatcher=all"
  traceFileName="trace.log"
/>
```

And well, I saw the message "the client doesn’t support push" – I used insomnia.

Then I started the frontend, logged in (to get a cookie/JWT), and opened the endpoint – vóila, it worked.

I wonder why no one ever wrote a blog post about this. 🤔 The referenced IBM blog post only mentions JSF, JSP and plain servlets. 

So, in the end it just worked fine! 👍🏻
